File: create-account/create-account.py
# アカウントをサービスカタログから作成する関数
# DynamoDB StreamによってトリガーされるLambda関数
# Lambda関数のevent変数には、DynamoDB Streamからのイベント情報が格納されている
# event変数の中身は以下のようになっている
# {
#   "Records": [
#     {
#       "eventID": "1",
#       "eventName": "INSERT",
#       "eventVersion": "1.0",
#       "eventSource": "aws:dynamodb",
#       "awsRegion": "us-west-2",
#       "dynamodb": {
#         "Keys": {
#           "email": {
#             "S": "メールアドレス"
#           }
#         },
#         "NewImage": {
#           "email": {
#             "S": "メールアドレス"
#           },
#           "first_name": {
#             "S": "名前"
#           },
#           "last_name": {
#             "S": "苗字"
#           },
#           "registration_date": {
#             "S": "登録日付"
#           }
#         },
#         "SequenceNumber": "111",
#         "SizeBytes": 26,
#         "StreamViewType": "NEW_AND_OLD_IMAGES"
#       },
#       "eventSourceARN": "eventsourcearn"
#     }
#   ]
# }
# この情報を使って、アカウントをサービスカタログに登録する
import boto3
import os

# AWS Budgetsを作成する関数
# 予算の基準を超えた場合に通知するSNSトピックを指定する
# 予算の基準は環境変数に設定されている
# 予算の基準は毎月の予算の半分を超えた場合と予算を超えた場合に通知する
# 予算はAWS Budgetsで設定する
def create_budget(account_id):
    # AWS Budgetsのクライアントを生成
    budgets = boto3.client('budgets')
    # 予算の基準を取得
    budget_limit = int(os.environ['BUDGET_LIMIT'])
    # 予算の基準を超えた場合に通知するSNSトピックを取得
    sns_topic = os.environ['SNS_TOPIC']
    # 予算を作成
    budgets.create_budget(
        AccountId=account_id,
        Budget={
            'BudgetName': 'MonthlyBudget',
            'BudgetLimit': budget_limit,
            'BudgetType': 'COST',
            'TimeUnit': 'MONTHLY',
            'TimePeriod': {
                'Start': '2000-01-01',
                'End': '2099-12-31'
            },
            'CalculatedSpend': {
                'ActualSpend': {
                    'Amount': '0',
                    'Unit': 'USD'
                }
            },
            'CostFilters': {
                'LinkedAccount': [account_id]
            },
            'CostTypes': {
                'IncludeTax': True,
                'IncludeSubscription': True,
                'UseBlended': True
            },
            'Notification': {
                'NotificationType': 'ACTUAL',
                'ComparisonOperator': 'GREATER_THAN',
                'Threshold': 100,
                'ThresholdType': 'PERCENTAGE',
                'NotificationState': 'ALARM'
            }
        }
    )
    # 予算の基準を超えた場合に通知する
    sns = boto3.client('sns')
    sns.publish(
        TopicArn=sns_topic,
        Message='The budget of account ' + account_id + ' has exceeded the limit'
    )
    # 予算の半分を超えた場合に通知する
    sns.publish(
        TopicArn=sns_topic,
        Message='The budget of account ' + account_id + ' has exceeded half of the limit'
    )

def lambda_handler(event,context):
    # service catalogからアカウント作成用のポートフォリオの情報を取得する。
    # この情報は環境変数に設定されている
    # デバッグ用にevent変数の中身を出力する
    product_id = os.environ['PRODUCT_ID']
    provisioning_artifact_id = os.environ['PROVISIONING_ARTIFACT_ID']
    managed_organizational_unit = os.environ['MANAGED_ORGANIZATIONAL_UNIT']
    email = event['Records'][0]['dynamodb']['NewImage']['email']['S']
    print(email)
    # プロビジョン後の製品の名前のために、emailから@以降を取り除いた文字列を取得する
    email_without_domain = email.split('@')[0]
    first_name = event['Records'][0]['dynamodb']['NewImage']['first_name']['S']
    last_name = event['Records'][0]['dynamodb']['NewImage']['last_name']['S']

    # service catalogのクライアントを生成する
    service_catalog = boto3.client('servicecatalog')
    # デバッグ用にservice catalogのプロダクト情報を取得
    print(service_catalog.list_provisioning_artifacts(ProductId=product_id))
    # service catalogの製品からアカウントを作成する
    service_catalog.provision_product(
        ProductId=product_id,
        ProvisioningArtifactId=provisioning_artifact_id,
        ProvisionedProductName=email_without_domain,
        ProvisioningParameters=[
            {
                'Key': 'AccountEmail',
                'Value': email
            },
            {
                'Key': 'AccountName',
                'Value': email_without_domain
            },
            {
                'Key': 'ManagedOrganizationalUnit',
                'Value': managed_organizational_unit
            },
            {
                'Key': 'SSOUserEmail',
                'Value': email
            },
            {
                'Key': 'SSOUserFirstName',
                'Value': first_name
            },
            {
                'Key': 'SSOUserLastName',
                'Value': last_name
            }
        ]
    )

File: script/create-account/create-account.go
package main

import (
	"context"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/aws/aws-lambda-go/lambda"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/budgets"
	"github.com/aws/aws-sdk-go/service/servicecatalog"
)

func createBudget(accountID string) {
	sess := session.Must(session.NewSession())
	budgetsClient := budgets.New(sess)

	budgetLimit := os.Getenv("BUDGET_LIMIT")
	snsTopic := os.Getenv("SNS_TOPIC")
	startDate, _ := time.Parse("2006-01-02", "2000-01-01")
	endDate, _ := time.Parse("2006-01-02", "2099-12-31")

	_, err := budgetsClient.CreateBudget(&budgets.CreateBudgetInput{
		AccountId: aws.String(accountID),
		Budget: &budgets.Budget{
			BudgetName: aws.String("MonthlyBudget"),
			BudgetLimit: &budgets.Spend{
				Amount: aws.String(budgetLimit),
				Unit:   aws.String("USD"),
			},
			BudgetType: aws.String("COST"),
			TimeUnit:   aws.String("MONTHLY"),
			TimePeriod: &budgets.TimePeriod{
				Start: &startDate,
				End:   &endDate,
			},
			CalculatedSpend: &budgets.CalculatedSpend{
				ActualSpend: &budgets.Spend{
					Amount: aws.String("0"),
					Unit:   aws.String("USD"),
				},
			},
			CostFilters: map[string][]*string{
				"LinkedAccount": {aws.String(accountID)},
			},
			CostTypes: &budgets.CostTypes{
				IncludeTax:          aws.Bool(true),
				IncludeSubscription: aws.Bool(true),
				UseBlended:          aws.Bool(true),
			},
		},
	})

	// Create notificationaa
	_, err = budgetsClient.CreateNotification(&budgets.CreateNotificationInput{
		AccountId:  aws.String(accountID),
		BudgetName: aws.String("MonthlyBudget"),
		Notification: &budgets.Notification{
			NotificationType:   aws.String("ACTUAL"),
			ComparisonOperator: aws.String("GREATER_THAN"),
			Threshold:          aws.Float64(100),
			ThresholdType:      aws.String("PERCENTAGE"),
			NotificationState:  aws.String("ALARM"),
		},
		Subscribers: []*budgets.Subscriber{
			{
				SubscriptionType: aws.String("SNS"),
				Address:          aws.String(snsTopic),
			},
		},
	})

	if err != nil {
		fmt.Println("Error creating notification:", err)
		return
	}

}

func lambdaHandler(ctx context.Context, event map[string]interface{}) {
	productID := os.Getenv("PRODUCT_ID")
	provisioningArtifactID := os.Getenv("PROVISIONING_ARTIFACT_ID")
	managedOrganizationalUnit := os.Getenv("MANAGED_ORGANIZATIONAL_UNIT")

	record := event["Records"].([]interface{})[0].(map[string]interface{})
	dynamodb := record["dynamodb"].(map[string]interface{})
	newImage := dynamodb["NewImage"].(map[string]interface{})
	email := newImage["email"].(map[string]interface{})["S"].(string)

	fmt.Println(email)

	emailWithoutDomain := strings.Split(email, "@")[0]
	firstName := newImage["first_name"].(map[string]interface{})["S"].(string)
	lastName := newImage["last_name"].(map[string]interface{})["S"].(string)

	sess := session.Must(session.NewSession())
	serviceCatalogClient := servicecatalog.New(sess)

	fmt.Println(serviceCatalogClient.ListProvisioningArtifacts(&servicecatalog.ListProvisioningArtifactsInput{
		ProductId: aws.String(productID),
	}))

	_, err := serviceCatalogClient.ProvisionProduct(&servicecatalog.ProvisionProductInput{
		ProductId:              aws.String(productID),
		ProvisioningArtifactId: aws.String(provisioningArtifactID),
		ProvisionedProductName: aws.String(emailWithoutDomain),
		ProvisioningParameters: []*servicecatalog.ProvisioningParameter{
			{
				Key:   aws.String("AccountEmail"),
				Value: aws.String(email),
			},
			{
				Key:   aws.String("AccountName"),
				Value: aws.String(emailWithoutDomain),
			},
			{
				Key:   aws.String("ManagedOrganizationalUnit"),
				Value: aws.String(managedOrganizationalUnit),
			},
			{
				Key:   aws.String("SSOUserEmail"),
				Value: aws.String(email),
			},
			{
				Key:   aws.String("SSOUserFirstName"),
				Value: aws.String(firstName),
			},
			{
				Key:   aws.String("SSOUserLastName"),
				Value: aws.String(lastName),
			},
		},
	})

	if err != nil {
		fmt.Println("Error provisioning product:", err)
		return
	}

	createBudget(emailWithoutDomain)
}

func main() {
	lambda.Start(lambdaHandler)
}
